lib/spack/spack/directory_layout.py

```diff
@@ -251,6 +251,50 @@ def build_env_path(self, spec):
     def build_packages_path(self, spec):
         return os.path.join(self.metadata_path(spec), self.packages_dir)
 
+    def _mkdirp(self, absolute_path, mode):
+        """
+        This mkdirp wrapper sets permissions to non-existing nested folders.
+
+        Nested folders are the ones which absolute_path prefix
+        is self.root (excluded)
+
+        e.g. absolute_path=self.root/a/b/c
+            If a path in set {a,b,c} does not exists, it will create it and
+            set permissions, otherwise it doesn't even touch it.
+            If self.root/a already exists, it will just create and
+            set permissions for b and c
+
+        Argument:
+            - absolute_path: (prefix must be be self.root)
+                absolute path of the folder hierarcy to create
+            - mode:
+                permissions to set to all non existing nested folders
+        """
+        assert(os.path.commonprefix([self.root, absolute_path]) == self.root)
+
+        nested_path = os.path.relpath(absolute_path, self.root)
+
+        nested_folders = []
```

Comment:
IMHO A more efficient approach -- assuming @tgamblin agrees -- is to change `mkdirp` to pass mode to `os.makedirs`.

Alternatively, if there is a solid rationale (or use cases) for _not_ using mode for intermediate directory permissions, then I'd be more inclined to add an option to `mkdirp` that allows `mode` to be passed to `os.makedirs`.

Reply:
FWIW. This was a comment I thought I'd added earlier, before Greg (becker33)'s response but I thought it worth "adding" anyway.

Reply:
Thank you for the comment @tldahlgren.

I opted for adding it to _directory_layout.py_ and not to _llnl/util/filesystem.py_ (as also @becker33 suggests) because it looked to me as a behavior tight to Spack directory layout management instead of a generic filesystem thing.

Moreover, at the beginning I thought I didn't want to alter the classic *mkdirp* function, but I agree with you that it may be a nice extension to have the option to set permissions for both intermediate and target folders, that may also be different.

Reply:
Supporting your suggestion, even `os.makedirs` in the past took into account permissions for intermediate directories (see [Python Issue19930](https://bugs.python.org/issue19930)), so I agree with you that we can extend `llnl/util/filesystem.py:mkdirs` to fit our needs.

In next commit, I'm going to add the option `mode_intermediate` to `mkdirs` function, so it will be possible to keep the same behavior as before (i.e. default OS permissions), but it will be also possible to set different permissions for newly-created folders and leaf folder.